// gulp-browserify-globs/index.js
'use strict';

var browserify = require('browserify');
var buffer = require('vinyl-buffer');
var globby = require('globby');
var objUtil = require('./lib/obj-util');
var source = require('vinyl-source-stream');
var sourcemaps = require('gulp-sourcemaps');
var through = require('through2');
var uglify = require('gulp-uglify');

var defaultOpts = {
  outfile: 'bundle.js',
  debug: true,
  transform: [],
  uglify: false,
  sourcemaps: false,
  sourcemapOut: './'
};

module.exports = function (globs, opts) {
  var bundledStream = through();

  opts = objUtil.extend(opts ? opts : {}, defaultOpts);

  globby(globs).then(function (entries) {
    var browserifyOpts = objUtil.without(opts, ['outfile', 'uglify']);
    browserifyOpts.entries = entries;

    var b = browserify(browserifyOpts);

    b.bundle().pipe(bundledStream);
  }).catch(function (err) {
    bundledStream.emit('error', err);
  });

  var stream = bundledStream
    .pipe(source(opts.outfile))
    .pipe(buffer());

  if (opts.sourcemaps) {
    var sourcemapOpts = typeof opts.sourcemaps === 'object' ? opts.sourcemaps : {loadMaps: true};
    stream = stream.pipe(sourcemaps.init(sourcemapOpts));
  }

  if (opts.uglify) {
    var uglifyOpts = typeof opts.uglify === 'object' ? opts.uglify : {};
    stream = stream.pipe(uglify(uglifyOpts));
  }

  if (opts.sourcemaps) {
    stream = stream.pipe(sourcemaps.write(opts.sourcemapOut));
  }

  return stream;
};
